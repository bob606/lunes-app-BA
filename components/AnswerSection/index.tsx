import {
  React,
  TouchableOpacity,
  useState,
  View,
  COLORS,
  styles,
  TextInput,
  Text,
  NextArrow,
  CloseIcon,
  IAnswerSectionProps,
  Popover,
  PopoverPlacement,
  PopoverContent,
  VolumeUpDisabled,
  InActiveVolumeUp,
  VolumeUp,
  Platform,
  SoundPlayer,
  Tts,
} from './imports';

const AnswerSection = ({
  count,
  index,
  setIndex,
  currentWordNumber,
  setCurrentWordNumber,
  document,
}: IAnswerSectionProps) => {
  const [word, setWord] = useState('');
  const touchable: any = React.useRef();
  const [isPopoverVisible, setIsPopoverVisible] = useState(false);
  const [isActive, setIsActive] = useState(false);
  const [isValidEntry, setIsValidEntry] = useState(false);

  const clearTextInput = () => {
    setWord('');
    setIsValidEntry(false);
  };

  const getNextWord = () => {
    if (index < count - 1) {
      setIndex(index + 1);
      setCurrentWordNumber(currentWordNumber + 1);
    }
  };

  const checkEntry = () => {
    if (word.trim().split(' ').length < 2) {
      setIsPopoverVisible(true);
      setIsValidEntry(false);
    } else {
      setIsValidEntry(true);
      setIsActive(true);
      handlaSpeakerClick(document?.audio);
    }
  };

  const handlaSpeakerClick = (audio?: string) => {
    setIsActive(true);

    // Don't use soundplayer for IOS, since IOS doesn't support .ogg files
    if (audio && Platform.OS !== 'ios') {
      //audio from API
      SoundPlayer.playUrl(document?.audio);
    } else {
      Tts.speak(document?.word, {
        androidParams: {
          KEY_PARAM_PAN: -1,
          KEY_PARAM_VOLUME: 0.5,
          KEY_PARAM_STREAM: 'STREAM_MUSIC',
        },
      });
    }
  };

  React.useEffect(() => {
    let _onSoundPlayerFinishPlaying: any = null;
    let _onTtsFinishPlaying: any = null;

    _onSoundPlayerFinishPlaying = SoundPlayer.addEventListener(
      'FinishedPlaying',
      () => {
        setIsActive(false);
      },
    );

    _onTtsFinishPlaying = Tts.addEventListener('tts-finish', () =>
      setIsActive(false),
    );

    return () => {
      _onSoundPlayerFinishPlaying.remove();
      _onTtsFinishPlaying.remove();
    };
  }, []);

  return (
    <View style={styles.container}>
      <Popover
        isVisible={isPopoverVisible}
        onRequestClose={() => setIsPopoverVisible(false)}
        from={touchable}
        placement={PopoverPlacement.TOP}
        arrowStyle={styles.arrow}
        arrowShift={-0.8}
        verticalOffset={-10}
        backgroundStyle={styles.overlay}>
        <PopoverContent />
      </Popover>

      {/* this is disabled for now, until finish implementing the logic */}
      <TouchableOpacity
        disabled={!word}
        style={styles.volumeIcon}
        onPress={() => handlaSpeakerClick(document?.audio)}>
        {isValidEntry && word ? (
          isActive ? (
            <VolumeUp />
          ) : (
            <InActiveVolumeUp />
          )
        ) : (
          <VolumeUpDisabled />
        )}
      </TouchableOpacity>

      <View
        ref={touchable}
        style={[styles.textInputContainer, !!word && styles.activeTextInput]}>
        <TextInput
          style={styles.textInput}
          placeholder="Enter Word with article"
          placeholderTextColor={COLORS.lunesBlackLight}
          value={word}
          onChangeText={(text) => setWord(text)}
        />
        {!!word && (
          <TouchableOpacity onPress={clearTextInput}>
            <CloseIcon />
          </TouchableOpacity>
        )}
      </View>

      <TouchableOpacity
        onPress={checkEntry}
        disabled={!word}
        style={[styles.checkEntryButton, !word && styles.disabledButton]}>
        <Text
          style={[styles.checkEntryLabel, !word && styles.disabledButtonLabel]}>
          Check entry
        </Text>
      </TouchableOpacity>

      <TouchableOpacity style={styles.giveUpButton}>
        <Text style={styles.giveUpLabel}>I give up!</Text>
      </TouchableOpacity>

      <TouchableOpacity style={styles.tryLaterButton} onPress={getNextWord}>
        <Text style={styles.giveUpLabel}>Try later</Text>
        <NextArrow />
      </TouchableOpacity>
    </View>
  );
};

export default AnswerSection;
