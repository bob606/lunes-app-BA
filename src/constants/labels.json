{
  "overview": {
    "welcome": "Willkommen bei Lunes!\nLerne Vokabeln für deinen Beruf.",
    "unit": "Lektion",
    "units": "Lektionen",
    "word": "Wort",
    "words": "Wörter",
    "exercises": "Übungen"
  },
  "header": {
    "overview": "Übersicht",
    "overviewExercises": "Übersicht der Übungen",
    "cancelExercise": "Übung beenden"
  },
  "exercises": {
    "overview": {
      "title": "Vokabelübersicht",
      "description": "Alle Wörte anschauen"
    },
    "wordSingleChoice": {
      "title": "Wort auswählen",
      "description": "Wähle das richtige Wort"
    },
    "articleSingleChoice": {
      "title": "Artikel üben",
      "description": "Wähle den richtigen Artikel"
    },
    "write": {
      "title": "Wort schreiben",
      "description": "Wort mit Artikel eingeben",
      "placeholder": {
        "insertAnswer": "Wort mit Artikel eingeben",
        "newTry": "Neuer Versuch"
      },
      "button": {
        "checkInput": "Eingabe überprüfen",
        "showSolution": "Lösung anzeigen",
        "tryLater": "Später versuchen"
      },
      "feedback": {
        "articleMissing": "Ups, deine Eingabe ist nicht vollständig. Füge einen Artikel ein.",
        "wrong": "Die richtige Antwort ist: ",
        "almostCorrect1": "Deine Eingabe ",
        "almostCorrect2": " ist fast richtig.",
        "correct": "Toll, weiter so! \nDeine Eingabe ist richtig."
      }
    },
    "buttons": {
      "next": "Nächstes Wort",
      "showResults": "Ergebnisse anzeigen"
    }
  },
  "results": {
    "resultsOverview": "Ergebnis-Übersicht",
    "feedbackGood": "Toll, weiter so! \nDu hast die Übung sehr gut gemeistert.",
    "feedbackMedium": "Übung macht den Meister, \nversuche es noch einmal!",
    "feedbackBad": "Nicht aufgeben! \nVersuche es noch einmal!",
    "checkEntries": "Eingaben überprüfen",
    "retryExercise": "Übung wiederholen",
    "wrong": "falsche",
    "similar": "ähnliche",
    "correct": "richtige",
    "entries": "Eingaben",
    "viewEntries": "Eingaben anschauen",
    "show": "Zeige",
    "of": "von"
  },
  "cancelModal": {
    "cancelAsk": "Möchtest du diese Übung wirklich beenden?",
    "cancel": "Beenden",
    "continue": "Weiter machen"
  }
}
