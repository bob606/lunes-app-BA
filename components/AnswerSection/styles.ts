import {StyleSheet} from 'react-native';
import {COLORS} from '../../constants/colors';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp,
} from 'react-native-responsive-screen';

export const styles = StyleSheet.create({
  container: {
    paddingVertical: 40,
    alignItems: 'center',
    position: 'relative',
    width: '100%',
    height: hp('65%'),
  },
  textInputContainer: {
    width: wp('80%'),
    height: hp('8%'),
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderWidth: 2,
    borderRadius: 2,
    paddingHorizontal: 10,
    marginBottom: hp('6%'),
  },
  textInput: {
    fontSize: 16,
    fontWeight: 'normal',
    letterSpacing: 0.11,
    fontFamily: 'SourceSansPro-Regular',
    color: COLORS.lunesBlack,
    width: wp('60%'),
  },
  volumeIcon: {
    position: 'absolute',
    top: -20,
    left: '45%',
    width: 40,
    height: 40,
    borderRadius: 50,
    backgroundColor: COLORS.lunesFunctionalIncorrectDark,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
