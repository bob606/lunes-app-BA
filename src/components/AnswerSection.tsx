import React, { useState } from 'react'
import { View, TouchableOpacity, TextInput, StyleSheet } from 'react-native'
import { CloseIcon } from '../../assets/images'
import { COLORS } from '../constants/colors'
import Popover from './Popover'
import Feedback from './FeedbackSection'
import stringSimilarity from 'string-similarity'
import Actions from './Actions'
import PopoverContent from './PopoverContent'
import { heightPercentageToDP as hp, widthPercentageToDP as wp } from 'react-native-responsive-screen'
import { DocumentType } from '../constants/endpoints'
import AsyncStorage from '../utils/AsyncStorage'
import { ARTICLES, ExerciseKeys, SimpleResultType } from '../constants/data'
import labels from '../constants/labels.json'
import AudioPlayer from './AudioPlayer'

export const styles = StyleSheet.create({
  container: {
    paddingVertical: 20,
    alignItems: 'center',
    position: 'relative',
    width: '100%',
    height: hp('65%')
  },
  textInputContainer: {
    width: wp('80%'),
    height: hp('8%'),
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 2,
    paddingHorizontal: 15,
    marginBottom: hp('6%')
  },
  textInput: {
    fontSize: wp('4.5%'),
    fontWeight: 'normal',
    letterSpacing: 0.11,
    fontFamily: 'SourceSansPro-Regular',
    color: COLORS.lunesBlack,
    width: wp('60%')
  }
})

export interface AnswerSectionPropsType {
  tryLater: () => void
  currentDocumentNumber: number
  setCurrentDocumentNumber: Function
  documents: DocumentType[]
  finishExercise: Function
  trainingSet: string
  disciplineTitle: string
}

const almostCorrectThreshold = 0.6

const AnswerSection = ({
  currentDocumentNumber,
  setCurrentDocumentNumber,
  finishExercise,
  tryLater,
  trainingSet,
  disciplineTitle,
  documents
}: AnswerSectionPropsType): JSX.Element => {
  const touchable: any = React.createRef()
  const [isPopoverVisible, setIsPopoverVisible] = useState(false)
  const [input, setInput] = useState('')
  const [submission, setSubmission] = useState('')
  const [result, setResult] = useState('')
  const [secondAttempt, setSecondAttempt] = useState(false)
  const document = documents[currentDocumentNumber]
  const totalNumbers = documents.length
  const [isFocused, setIsFocused] = useState(false)

  function capitalizeFirstLetter(string: string) {
    return string.charAt(0).toUpperCase() + string.slice(1)
  }

  const checkEntry = (): void => {
    setSubmission(input)
    const splitInput = input.trim().split(' ')

    if (splitInput.length < 2) {
      setIsPopoverVisible(true)
      return
    }

    const article = capitalizeFirstLetter(splitInput[0])
    const word = splitInput[1]

    if (!validateForSimilar(article, word)) {
      setResult('incorrect')
      storeResult('incorrect')
    } else if (validateForCorrect(article, word)) {
      setResult('correct')
      storeResult('correct')
    } else if (secondAttempt) {
      setResult('similar')
      storeResult('similar')
    } else {
      setInput('')
      setSecondAttempt(true)
      return
    }
    setSecondAttempt(false)
  }

  const validateForCorrect = (inputArticle: string, inputWord: string): boolean => {
    const exactAnswer = inputArticle === document?.article.value && inputWord === document?.word

    const altAnswer = document?.alternatives?.some(
      ({ article, alt_word: altWord }) => inputArticle === ARTICLES[article].value && inputWord === altWord
    )
    return exactAnswer || altAnswer
  }

  const validateForSimilar = (inputArticle: string, inputWord: string): boolean => {
    if (validateForCorrectWithoutArticle(inputWord)) {
      return true
    }
    const origCheck = stringSimilarity.compareTwoStrings(inputWord, document.word) > almostCorrectThreshold

    const altCheck = document.alternatives.some(
      ({ article, alt_word: altWord }) =>
        inputArticle === ARTICLES[article].value &&
        stringSimilarity.compareTwoStrings(inputWord, altWord) > almostCorrectThreshold
    )
    return origCheck || altCheck
  }

  const giveUp = (): void => {
    setResult('giveUp')
    storeResult(secondAttempt ? 'similar' : 'incorrect')
  }

  const validateForCorrectWithoutArticle = (inputWord: string): boolean => {
    const exactAnswer = inputWord === document?.word

    const altAnswer = document?.alternatives?.some(({ article, alt_word: altWord }) => inputWord === altWord)
    return exactAnswer || altAnswer
  }

  const getNextWord = (): void => {
    setResult('')
    setInput('')
    setSecondAttempt(false)

    if (currentDocumentNumber === totalNumbers - 1) {
      finishExercise()
      return
    }
    setCurrentDocumentNumber(currentDocumentNumber + 1)
  }

  const storeResult = async (score: SimpleResultType): Promise<void> => {
    try {
      const exercise = (await AsyncStorage.getExercise(ExerciseKeys.writeExercise)) ?? {}
      exercise[disciplineTitle] = exercise[disciplineTitle] ?? {}
      exercise[disciplineTitle][trainingSet] = exercise[disciplineTitle][trainingSet] ?? {}

      exercise[disciplineTitle][trainingSet][document.word] = {
        ...document,
        result: score
      }

      await AsyncStorage.setExercise(ExerciseKeys.writeExercise, exercise)

      const session = await AsyncStorage.getSession()
      if (session === null) {
        throw new Error('Session is not saved correctly!')
      }
      const newSession = {
        ...session,
        retryData: {
          data: documents.slice(currentDocumentNumber + 1, totalNumbers)
        }
      }
      await AsyncStorage.setSession(newSession)
    } catch (e) {
      console.error(e)
    }
  }

  const getBorderColor = (): string => {
    if (isFocused) {
      return COLORS.lunesBlack
    } else if (!secondAttempt && !input) {
      return COLORS.lunesGreyMedium
    } else if (!result && !secondAttempt) {
      return COLORS.lunesBlack
    } else if (result === 'correct') {
      return COLORS.lunesFunctionalCorrectDark
    } else if (result === 'incorrect' || !secondAttempt) {
      return COLORS.lunesFunctionalIncorrectDark
    } else {
      return COLORS.lunesFunctionalAlmostCorrectDark
    }
  }

  return (
    <>
      <AudioPlayer document={document} disabled={result === '' && !secondAttempt} />
      <View style={styles.container}>
        <Popover isVisible={isPopoverVisible} setIsPopoverVisible={setIsPopoverVisible} ref={touchable}>
          <PopoverContent />
        </Popover>

        <View
          testID='input-field'
          ref={touchable}
          style={[
            styles.textInputContainer,
            {
              borderColor: getBorderColor()
            }
          ]}>
          <TextInput
            style={styles.textInput}
            placeholder={secondAttempt ? labels.exercises.write.newTry : labels.exercises.write.insertAnswer}
            placeholderTextColor={COLORS.lunesBlackLight}
            value={input}
            onChangeText={text => setInput(text)}
            editable={result === ''}
            onFocus={() => setIsFocused(true)}
            onBlur={() => setIsFocused(false)}
          />
          {(isFocused || (result === '' && input !== '')) && (
            <TouchableOpacity onPress={() => setInput('')}>
              <CloseIcon />
            </TouchableOpacity>
          )}
        </View>

        <Feedback secondAttempt={secondAttempt} result={result} document={document} input={submission} />

        <Actions
          tryLater={tryLater}
          giveUp={giveUp}
          input={input}
          result={result}
          checkEntry={checkEntry}
          getNextWord={getNextWord}
          secondAttempt={secondAttempt}
          isFinished={currentDocumentNumber === totalNumbers - 1}
        />
      </View>
    </>
  )
}

export default AnswerSection
