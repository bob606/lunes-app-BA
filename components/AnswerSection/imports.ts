import React, {useState} from 'react';
import {View, TouchableOpacity, TextInput, Platform} from 'react-native';
import {styles} from './styles';
import {CloseIcon, VolumeUp} from '../../assets/images/imports';
import {COLORS} from '../../constants/colors';
import {IAnswerSectionProps, IDocumentProps} from '../../interfaces/exercise';
import Popover from '../Popover';
import SoundPlayer from 'react-native-sound-player';
import Tts from 'react-native-tts';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Feedback from '../FeedbackSection';
import stringSimilarity from 'string-similarity';
import Actions from '../Actions';
import {SCREENS} from '../../constants/data';
import PopoverContent from '../PopoverContent';
import {useFocusEffect} from '@react-navigation/native';

export {
  React,
  View,
  styles,
  TouchableOpacity,
  useState,
  COLORS,
  TextInput,
  CloseIcon,
  Popover,
  VolumeUp,
  Platform,
  SoundPlayer,
  Tts,
  AsyncStorage,
  Feedback,
  stringSimilarity,
  Actions,
  SCREENS,
  PopoverContent,
  useFocusEffect,
};

export type {IAnswerSectionProps, IDocumentProps};
