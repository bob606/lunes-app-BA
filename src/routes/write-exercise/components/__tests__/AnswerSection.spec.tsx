import React from 'react'
import AnswerSection, { AnswerSectionPropsType } from '../AnswerSection'
import { fireEvent, render } from '@testing-library/react-native'
import labels from '../../../../constants/labels.json'

jest.mock('../../../../components/AudioPlayer', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const Text = require('react-native').Text
  return () => <Text>AudioPlayer</Text>
})

jest.mock('react-native/Libraries/LogBox/Data/LogBoxData')

describe('AnswerSection', () => {
  const defaultAnswerSectionProps: AnswerSectionPropsType = {
    documents: [
      {
        id: 1,
        word: 'Spachtel',
        article: {
          id: 1,
          value: 'Der'
        },
        document_image: [{ id: 1, image: '' }],
        audio: '',
        alternatives: [
          {
            word: 'Spachtel',
            article: {
              id: 2,
              value: 'Die'
            }
          },
          {
            word: 'Alternative',
            article: {
              id: 2,
              value: 'Die'
            }
          }
        ]
      }
    ],
    trainingSet: '',
    disciplineTitle: '',
    currentDocumentNumber: 0,
    tryLater: () => {},
    finishExercise: () => {},
    setCurrentDocumentNumber: () => {}
  }

  const evaluate = async (input: string, expectedFeedback: string) => {
    const { getByPlaceholderText, getByText, getByTestId } = render(<AnswerSection {...defaultAnswerSectionProps} />)
    const inputField = await getByPlaceholderText(labels.exercises.write.insertAnswer)
    await fireEvent.changeText(inputField, input)
    const button = await getByText(labels.exercises.write.checkInput)
    await fireEvent.press(button)
    const result = await getByTestId('feedback-write-exercise')
    expect(result.children[0]).toEqual(expectedFeedback)
  }

  it('should show correct-feedback for correct solution', async () => {
    await evaluate('Der Spachtel', labels.exercises.write.feedback.correct)
  })

  it('should show correct-feedback for alternative article', async () => {
    await evaluate('Die Spachtel', labels.exercises.write.feedback.correct)
  })

  it('should show correct-feedback for alternative solution', async () => {
    await evaluate('Die Alternative', labels.exercises.write.feedback.correct)
  })

  it('should show almost correct feedback', async () => {
    const input = 'Die Spachtl'
    await evaluate(
      input,
      `${labels.exercises.write.feedback.almostCorrect1} „${input}“ ${labels.exercises.write.feedback.almostCorrect2}`
    )
  })

  it('should show wrong feedback', async () => {
    evaluate(
      'Das Falsche',
      `${labels.exercises.write.feedback.wrong} ${defaultAnswerSectionProps.documents[0].article.value} ${defaultAnswerSectionProps.documents[0].word}`
    )
  })
})
