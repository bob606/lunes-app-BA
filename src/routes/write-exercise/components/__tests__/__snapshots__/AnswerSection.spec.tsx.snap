// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerSection renders correctly across screens 1`] = `
<Fragment>
  <AudioPlayer
    disabled={true}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 867,
        "paddingVertical": 20,
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <ForwardRef
      isVisible={false}
      setIsPopoverVisible={[Function]}
    >
      <PopoverContent />
    </ForwardRef>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 2,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 106.5,
            "justifyContent": "space-between",
            "marginBottom": 80,
            "paddingHorizontal": 15,
            "width": 600,
          },
          Object {
            "borderColor": "#5b616a",
          },
        ]
      }
      testID="input-field"
    >
      <Component
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Wort mit Artikel eingeben"
        placeholderTextColor="#777e91"
        rejectResponderTermination={true}
        style={
          Object {
            "color": "#000e38",
            "fontFamily": "SourceSansPro-Regular",
            "fontSize": 34,
            "fontWeight": "normal",
            "letterSpacing": 0.11,
            "width": 450,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Feedback
      input=""
      result=""
      secondAttempt={false}
    />
    <Actions
      checkEntry={[Function]}
      getNextWord={[Function]}
      giveUp={[Function]}
      input=""
      isFinished={false}
      result=""
      secondAttempt={false}
      tryLater={[Function]}
    />
  </View>
</Fragment>
`;
